Updates for TauReco at miniAOD

mbluj opened the pull request:
#### PR description:

This PR contains updates for Tau reconstruction on top of MiniAOD. 
The main change is addition to the miniAOD event content energy fractions of PFGammas (packedCandidates with pdgId==22). This completes information needed to build tau quantities based on energy fractions of its constituents. Other changes introduced by the PR are use of the energy fractions:
* updates of tauID against-mu simple: usage of previously added energy fractions for charged hadrons and code reorganisation to use new tauID dataformat + fillDescription;
* anti-e MVA6 and deepTauID (both using energy fractions of charged hadrons and gammas) added to the tauReco-at-MiniAOD workflow;
* recently moderinised tauID against-e-in-deadECal added back to the tauReco-at-MiniAOD workflow;

#### PR validation:
Matrix tests (`runTheMatrix.py -l limited -i all --ibeos`) successful.

mbluj committed:
Store energy fractions for gammas, and then use to build pat::taus on top of MiniAOD

mbluj committed:
Switch to corrected energy fraction and enable calo veto in simple anti-mu tauID

mbluj committed:
Code quality fixes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/16808

- This PR adds an extra 48KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc  modified in PR(s): #30508, #29927, #29933, #29935, #29936
  - File  PhysicsTools/PatAlgos/plugins/PATTauProducer.cc  modified in PR(s): #29935, #29936, #29937
  - File  RecoTauTag/RecoTau/src/PositionAtECalEntranceComputer.cc  modified in PR(s): #30508

cmsbuild commented:
A new Pull Request was created by  @mbluj  for master.

It involves the following packages:

PhysicsTools/PatAlgos
RecoTauTag/Configuration
RecoTauTag/RecoTau


@perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @gouskos, @hatakeyamak, @emilbols, @peruzzim, @seemasharmafnal, @mmarionncern, @ahinzmann, @smoortga, @jdolen, @ferencek, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @andrzejnovak, @clelange, @riga, @JyothsnaKomaragiri, @mverzett, @gpetruc, @mariadalfonso this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 341f521ab6a67904a4706c42d094765591d45975
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c46113/7749/summary.html
CMSSW: CMSSW_11_2_X_2020-07-06-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c46113/7749/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 77 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2787364
 - DQMHistoTests: Total failures: 14
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2787299
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 154 log files, 17 edm output root files, 37 DQM output files

mbluj committed:
Run anti-e and deepTau on top of taus@mini

santocch commented:
+1

slava77 commented on RecoTauTag/RecoTau/python/hpsPFTauDiscriminationByAMuonRejectionSimple_cff.py:
```diff
@@ -3,12 +3,12 @@
-    srcPatMuons=cms.InputTag("slimmedMuons"),
-    maxNumberOfSTAMuons=cms.int32(-1),
-    maxNumberOfRPCMuons=cms.int32(-1),
+    PFTauProducer = cms.InputTag("hpsPFTauProducer"),
+    HoPMin = hpsPFTauDiscriminationByMuonRejection3.IDWPdefinitions[0].HoPMin,
+    doCaloMuonVeto = cms.bool(True),
+    srcPatMuons = cms.InputTag("slimmedMuons"),
+    maxNumberOfSTAMuons = cms.int32(-1),
+    maxNumberOfRPCMuons = cms.int32(-1),
```
please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.
```suggestion
    PFTauProducer = "hpsPFTauProducer",
    HoPMin = hpsPFTauDiscriminationByMuonRejection3.IDWPdefinitions[0].HoPMin,
    doCaloMuonVeto = True,
    srcPatMuons = "slimmedMuons",
    maxNumberOfSTAMuons = -1,
    maxNumberOfRPCMuons = -1,
```

slava77 commented:
assign xpog
to sign off on the miniAOD output size increase

> > The main change is addition to the miniAOD event content energy fractions of PFGammas (packedCandidates with pdgId==22).
> 
> please provide event size cost information of this update

in 100 events of wf 136.88811 available in jenkins outputs
the packedCandidates size is up by 83 bytes/event or 0.1% of the file size.

1K events or more would be a safer check though, although I doubt that it would change the fraction to above 0.2%

cmsbuild commented:
New categories assigned: xpog

@fgolf,@mariadalfonso,@gouskos,@peruzzim you have been requested to review this Pull request/Issue and eventually sign? Thanks

slava77 reviewed (commented):
the comment affects only slightly reformatted part of what was already in the code.
We'll likely address it in the general migration/update PRs.
So, I'm not holding the reco signature for this one (feel free to ignore or postpone for another update).

slava77 commented:
> Other changes introduced by the PR are use of the energy fractions.

comparisons show small changes in deepTau quantities.
(I didn't check explicitly) I suppose that if the photon inputs now have hcal and ecal energies available, this can suddenly become a significant shift in the discriminating power.
Was it validated, or do we just assume that there will be a retraining anyways?

mbluj commented:
I checked that increase of output file size is 0.1% using a ttbar relval sample, but I can redo the test. Could you remind me if there are tools to check increase of branch size?
I will also implement fixes as proposed in above comments.

mbluj commented:
> 
> 
> > Other changes introduced by the PR are use of the energy fractions.
> 
> comparisons show small changes in deepTau quantities.
> (I didn't check explicitly) I suppose that if the photon inputs now have hcal and ecal energies available, this can suddenly become a significant shift in the discriminating power.
> Was it validated, or do we just assume that there will be a retraining anyways?

No, it was not validatd as it was assumed that not filled fractions, i.e. ones equal to zero are not used. I will check it and eventually fix for backward compatibility. But, a new training is indeed expected for run-3.

perrotta commented:
> I checked that increase of output file size is 0.1% using a ttbar relval sample, but I can redo the test. Could you remind me if there are tools to check increase of branch size?
> I will also implement fixes as proposed in above comments.

@mbluj a tool that we use in reco to compare the size of output products  is https://twiki.cern.ch/twiki/bin/view/CMS/RecoIntegration#Compare_output_products

mbluj committed:
drop parameter type specifications in a module clonning

mbluj commented:
Size increase for MiniAOD on top of 9k ttbar relval Run-3 events [1] with the updated setup compared to plain CMSSW_11_2_0_pre1; configuration with cmsDriver command specified in [2].
* total size: 702258766/701292254 bytes = 1.0014, i.e. +0.14% or +107 bytes/event,
* packedPFCandidates: +0.3% (output from compareProducts.sh [3]).

[1] Sample taken from `root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_11_2_0_pre1/RelValTTbar_14TeV/GEN-SIM-RECO/`
[2] `cmsDriver miniAOD2021 --filein root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_11_2_0_pre1/RelValTTbar_14TeV/GEN-SIM-RECO/PU_111X_mcRun3_2021_realistic_v4-v2/10000/04D4491C-E468-7D44-AB1F-A9350C7E4A76.root --fileout file:MiniAOD2021.root --mc --eventcontent MINIAODSIM --runUnscheduled --datatier MINIAODSIM --conditions auto:phase1_2021_realistic --step PAT --nThreads 4 --geometry DB:Extended --scenario pp --era Run3 -n -1 --no_exec`
[3] `33364.2 ->     33473.6        109      0.3   0.15     patPackedCandidates_packedPFCandidates__PAT.`

mbluj commented:
To keep you informed about deepTau related checks:
* There are indeed differences in deepTau scores obtained with original and updated CMSSW, which however does not change significantly shape of distributions of the scores. It means that it is expected that global performances do not change despite differences at single tau basis;
* DeepTau does not use energy fractions neither of pf-gammas nor pf-charged (the latter were also introduced after deepTau has been trained); the only used fractions are for isolated charged hadrons and neutral hadrons. Therefore, I suppose that the difference is caused by some indirect dependence - I am checking it.

mbluj committed:
check track presence before computing energy sums (reverts to previous behaviour)

cmsbuild commented:
The code-checks are being triggered in jenkins.

mbluj commented:
Recent three commits do the following:
* removes type specification when module cloned in a configuration file (fix requested during review),
* fixes issue with different deepTau scores,
* adds anti-e and deepTau to a test workflow of tau reco at Mini.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/17066

- This PR adds an extra 48KB to repository

cmsbuild commented:
Pull request #30571 was updated. @perrotta, @gouskos, @cmsbuild, @fgolf, @slava77, @jpata, @mariadalfonso, @santocch, @peruzzim can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 459440fdc4d9c0fbebd57e931f8f71dc449dda61

CMSSW: CMSSW_11_2_X_2020-07-16-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c46113/8006/summary.html

I found follow errors while testing this PR

**Failed tests:  Python3**

* **Python3**:

I found errors: 
 
 *** Error compiling 'src/RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py'...

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

slava77 commented:
> *** Error compiling 'src/RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py'...

these are like
> SyntaxWarning: "is not" with a literal. Did you mean "!="?

mbluj commented:
> 
> 
> > *** Error compiling 'src/RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py'...
> 
> these are like
> 
> > SyntaxWarning: "is not" with a literal. Did you mean "!="?

Actually, it is because type specifiers are not present when cloning anti-mu discrimiantns. It is a the cloning is firstly preformed for one plugin (EDProducer) type and then type of the plugin is changed. The two types of modules have a bit different sets of arguments. So for arguments present in the new type but not in the old one type full argument type specification is needed. The previous way of doing was a "hack" which was forgotten. A clean solution requires some coding: one should add fillDescriptions to the new module and make it compliant with the new tauID dataformat. I will try to fix it by this weekend.

slava77 commented:
> A clean solution requires some coding: one should add fillDescriptions to the new module and make it compliant with the new tauID dataformat. I will try to fix it by this weekend.

@mbluj 
please clarify on the status of this.
Thank you.

mbluj commented:
> 
> 
> > A clean solution requires some coding: one should add fillDescriptions to the new module and make it compliant with the new tauID dataformat. I will try to fix it by this weekend.
> 
> @mbluj
> please clarify on the status of this.
> Thank you.

Unfortunately, I was not able to finish it in expected time. This and next week I am off, but I try to find some time to finish this development.

slava77 commented:
> Unfortunately, I was not able to finish it in expected time. This and next week I am off, but I try to find some time to finish this development.

ping

slava77 commented:
> > Unfortunately, I was not able to finish it in expected time. This and next week I am off, but I try to find some time to finish this development.
> 
> ping

slava77 commented:
> > > Unfortunately, I was not able to finish it in expected time. This and next week I am off, but I try to find some time to finish this development.
> > 
> > 
> > ping

@mbluj 
(just in case a message without a tag is less noticeable)

mbluj commented:
> 
> 
> > > > Unfortunately, I was not able to finish it in expected time. This and next week I am off, but I try to find some time to finish this development.
> > > 
> > > 
> > > ping
> 
> @mbluj
> (just in case a message without a tag is less noticeable)

This PR is not forgotten, but I spend my time for development working on various UL re-miniAOD related tau stuff. I hope go back to it soon.

mbluj committed:
Modernise anti-mu simple discriminant, adapt tauReco-at-mini

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/17708

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/17708/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/17708/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/17708/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/17708/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

mbluj committed:
Code checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/17709

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #30571 was updated. @perrotta, @gouskos, @cmsbuild, @fgolf, @slava77, @jpata, @mariadalfonso, @santocch, @peruzzim can you please check and sign again.

mbluj commented:
Recent commits update anti-mu simple tauID, adds back anti-e-in-deadECal tauID (possible thanks to its modernization in #29627). 
In addition, development branch of the PR is rebased to a newer development release.
PR description updated to provide more precise information on its current content.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 54785f1889e9448c742f2cc21294c21fa84146b5

CMSSW: CMSSW_11_2_X_2020-08-12-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c46113/8727/summary.html

I found follow errors while testing this PR

**Failed tests:  Python3**

* **Python3**:

I found errors: 
 
 *** Error compiling 'src/RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py'...

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

mbluj commented:
> 
> 
> Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

Checking it, but noting incorrect with `RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py` found in my earlier tests...

mbluj committed:
indentation fix

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30571/17715

- This PR adds an extra 48KB to repository

cmsbuild commented:
Pull request #30571 was updated. @perrotta, @gouskos, @cmsbuild, @fgolf, @slava77, @jpata, @mariadalfonso, @santocch, @peruzzim can you please check and sign again.

mbluj commented:
Indentation fixed. What is strange is that issue with indentation is neither seen by eye with emacs, which apparently does it differently than other tool which must been previously used, nor does appear with local unit tests.
BTW, could indentation issues of this kind ("TabError: inconsistent use of tabs and spaces in indentation") be automatically detected with code-format or code-style (or similar) checks?

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
>  BTW, could indentation issues of this kind ("TabError: inconsistent use of tabs and spaces in indentation") be automatically detected with code-format or code-style (or similar) checks?

I opened #30775 a few weeks ago but there was no traction yet

cmsbuild commented:
+1
Tested at: 7b7ad325caaefbc17d91cde6473e099f319b2338
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c46113/8728/summary.html
CMSSW: CMSSW_11_2_X_2020-08-12-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c46113/8728/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 25 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612401
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2612347
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

slava77 commented:
> DeepTau does not use energy fractions neither of pf-gammas nor pf-charged (the latter were also introduced after deepTau has been trained); the only used fractions are for isolated charged hadrons and neutral hadrons. Therefore, I suppose that the difference is caused by some indirect dependence - I am checking it.

@mbluj 
please clarify if the check was done and did not show any issues to be fixed in this PR
Thank you.

slava77 commented:
> > DeepTau does not use energy fractions neither of pf-gammas nor pf-charged (the latter were also introduced after deepTau has been trained); the only used fractions are for isolated charged hadrons and neutral hadrons. Therefore, I suppose that the difference is caused by some indirect dependence - I am checking it.
> 
> @mbluj
> please clarify if the check was done and did not show any issues to be fixed in this PR
> Thank you.

nevermind, this PR does not modify deepTau values anymore

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/30571	7b7ad32 
- code changes are in line with the PR description and the follow up review
    - general miniAOD is updated to have hcalFraction and caloFraction saved for photons (pt>0.5 GeV); this is estimated to increase the size of packedPFCandidates by 0.3% https://github.com/cms-sw/cmssw/pull/30571#issuecomment-659328568
    - other changes are affecting tau reco from miniAOD, which is not in the cms-bot tests
- jenkins tests pass and comparisons with the baseline show differences only in the packedPFCandidates caloFraction (curiously, in with the statistics of the tests there were no changes in the hcalFraction variable)

mbluj commented:
> 
> 
> > > DeepTau does not use energy fractions neither of pf-gammas nor pf-charged (the latter were also introduced after deepTau has been trained); the only used fractions are for isolated charged hadrons and neutral hadrons. Therefore, I suppose that the difference is caused by some indirect dependence - I am checking it.
> > 
> > 
> > @mbluj
> > please clarify if the check was done and did not show any issues to be fixed in this PR
> > Thank you.
> 
> nevermind, this PR does not modify deepTau values anymore

Yes indeed an issue affecting deepTauID (not diectly related with the energy fractions) has been introduced in this PR, and then fixed.

santocch commented:
+1

slava77 commented:
@cms-sw/xpog-l2 
please check this PR, your signature is needed.
Thank you.

mariadalfonso commented:
+xpog
0.3% increase for packedPFCandidates is acceptable

> Size increase for MiniAOD on top of 9k ttbar relval Run-3 events [1] with the updated setup compared to plain CMSSW_11_2_0_pre1; configuration with cmsDriver command specified in [2].
> 
> * total size: 702258766/701292254 bytes = 1.0014, i.e. +0.14% or +107 bytes/event,
> * packedPFCandidates: +0.3% (output from compareProducts.sh [3]).
>

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
